### PR TITLE
User two factor authentication

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
This documentation is related to the new feature User two factor authentication in CloudStack.

The corresponding PR is https://github.com/apache/cloudstack/pull/6924

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by weizhouapache** [comment]
@blueorangutan docbuild

**Comment by weizhouapache** [comment]
@blueorangutan docbuild

**Comment by weizhouapache** [comment]
doc looks good to me
https://qa.cloudstack.cloud/builds/docs-build/pr/293/adminguide/accounts.html#using-two-factor-authentication-for-users

**Review by weizhouapache — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
please check https://qa.cloudstack.cloud/builds/docs-build/pr/293/adminguide/accounts.html#using-two-factor-authentication-for-users

File: source/adminguide/accounts.rst [hunk 1]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
 
 - ``saml2.timeout``: SAML2 IDP Metadata refresh interval in seconds, minimum value is set to 300. Default is 1800
 
+Using Two Factor Authentication For Users
+------------------------------------------
+
+CloudStack supports two factor authentication 2FA for the users, wherein users needs
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:633]:
```suggestion
CloudStack supports two factor authentication, wherein users need
```

File: source/adminguide/accounts.rst [hunk 2]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+to provide a 2FA code after the login using username and password. CloudStack currently
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:634]:
```suggestion
to provide a 2FA code after the regular login using username and password. CloudStack currently
```

File: source/adminguide/accounts.rst [hunk 3]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+supports google authenticator along with other TOTP authenticators and staticpin as the 2FA providers.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:635]:
```suggestion
supports Google Authenticator or other TOTP authenticators and static PIN as the 2FA providers.
```

File: source/adminguide/accounts.rst [hunk 4]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+Other 2FA providers can be easily integrated with CloudStack using its plugin model.
+
+For admins, following are the settings available at global and domain level to configure
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:638]:
```suggestion
For admins, the following are the settings available at global and domain level to configure
```

File: source/adminguide/accounts.rst [hunk 5]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+   :alt: Verify 2FA page
+
+Users can configure 2FA in CloudStack using the action button in user form.
+
+.. image:: /_static/images/configure-2fa-action-button.png
+   :width: 400px
+   :align: center
+   :alt: Configure 2FA action button
+
+In the 2FA setup form, user needs to select one of the providers. CloudStack currently supports
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:665]:
```suggestion
In the 2FA setup form, the user needs to select one of the providers. CloudStack currently supports
```

File: source/adminguide/accounts.rst [hunk 6]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+google authenticator along with other TOTP authenticators and staticpin as the 2FA providers.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:666]:
```suggestion
Google Authenticator or other TOTP authenticators and static PIN as the 2FA providers.
```

File: source/adminguide/accounts.rst [hunk 7]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+When the google or other TOTP 2FA provider is selected, user must setup the account in google or other TOTP authenticator
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:668]:
```suggestion
When the Google Authenticator or Other TOTP 2FA provider is selected, the user must setup the account in the respective 
```

File: source/adminguide/accounts.rst [hunk 8]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+application in their device by either scanning the QR code or using the setup key provided by CloudStack.
+Once the account is setup in the authenticator application, user must verify 2FA with Cloudstack using
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:670]:
```suggestion
Once the account is setup in the authenticator application, the user must verify 2FA with Cloudstack using
```

File: source/adminguide/accounts.rst [hunk 9]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+the 2FA code in the authenticator application. Every time user has to use the same account in the
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:671]:
```suggestion
the 2FA code in the authenticator application. Every time the user has to use the same account in the
```

File: source/adminguide/accounts.rst [hunk 10]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+authenticator application during login into CloudStack.
+
+.. image:: /_static/images/configure-google-2fa-form.png
+   :width: 400px
+   :align: center
+   :alt: Configure google 2FA form
+
+When the staticpin 2FA provider is selected, user must use the static pin as the 2FA code to verify 2FA
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:679]:
```suggestion
When the static PIN 2FA provider is selected, the user must use the static PIN as the code to verify 2FA
```

File: source/adminguide/accounts.rst [hunk 11]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+with CloudStack. Every time user has to use this static pin as 2FA code during login into CloudStack.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:680]:
```suggestion
with CloudStack. The user must input this static PIN as a 2FA code every time they need to login.
```

File: source/adminguide/accounts.rst [hunk 12]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+.. image:: /_static/images/configure-staticpin-2fa-form.png
+   :width: 400px
+   :align: center
+   :alt: Configure staticpin 2FA form
+
+.. Note:: While user configures 2FA, verification of the 2FA code is must to complete the process.
+
+Admin has the capability to mandate 2FA for users.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:689]:
```suggestion
The admin has the capability to mandate 2FA for users.
```

File: source/adminguide/accounts.rst [hunk 13]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+When a new user is created ``mandate.user.two.factor.authentication`` setting will be used to enable
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:690]:
```suggestion
The setting ``mandate.user.2fa`` is used to enable
```

File: source/adminguide/accounts.rst [hunk 14]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+2FA for the user. In that case user must configure 2FA during the first login into CloudStack.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:691]:
```suggestion
2FA for the user. In that case the user must configure 2FA during their first login into CloudStack.
```

File: source/adminguide/accounts.rst [hunk 15]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+User's first login page to configure 2FA looks like below,
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:692]:
```suggestion
The user's first login page to configure 2FA looks like the below,
```

**Review by NuxRo — CHANGES_REQUESTED** [review]
Minor edits for correctness

File: source/adminguide/accounts.rst [hunk 16]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+   :width: 400px
+   :align: center
+   :alt: Configure 2FA action button
+
+In the 2FA setup form, the user needs to select one of the providers. CloudStack currently supports
+Google Authenticator or other TOTP Authenticators and static PIN as the 2FA providers.
+
+When the Google Authenticator or other TOTP 2FA provider is selected, the user must setup the account in
+the respective application in their device by either scanning the QR code or using the setup key provided
+by CloudStack. Once the account is setup in the authenticator application, the user must verify 2FA with
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:669]:
```suggestion
by CloudStack. Once this is set up in the authenticator application, the user must always use the provided 2FA codes to log in.
```

File: source/adminguide/accounts.rst [hunk 17]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+Cloudstack using the 2FA code in the authenticator application. Every time the user has to use the same
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:670]:
```suggestion

```

File: source/adminguide/accounts.rst [hunk 18]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+account in the authenticator application during login into CloudStack.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:671]:
```suggestion
```

File: source/adminguide/accounts.rst [hunk 19]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+.. image:: /_static/images/configure-google-2fa-form.png
+   :width: 400px
+   :align: center
+   :alt: Configure google 2FA form
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:676]:
```suggestion
   :alt: Configure Google 2FA form
```

File: source/adminguide/accounts.rst [hunk 20]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+When the static PIN 2FA provider is selected, the user must use the static PIN as the code to verify 2FA
+with CloudStack. The user must input this static PIN as a 2FA code every time they need to login.
+
+.. image:: /_static/images/configure-staticpin-2fa-form.png
+   :width: 400px
+   :align: center
+   :alt: Configure staticpin 2FA form
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:684]:
```suggestion
   :alt: Configure static PIN 2FA form
```

File: source/adminguide/accounts.rst [hunk 21]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+.. Note:: While user configures 2FA, verification of the 2FA code is must to complete the process.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:686]:
```suggestion
.. Note:: While the user configures 2FA, verification of the 2FA code is a must to complete the process.
```

File: source/adminguide/accounts.rst [hunk 22]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+The admin has the capability to mandate 2FA for users. The setting ``mandate.user.2fa``
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:688]:
```suggestion
The admin has the capability to mandate 2FA for users via the setting ``mandate.user.2fa``.
```

File: source/adminguide/accounts.rst [hunk 23]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+is used to enable 2FA for the user. In that case the user must configure 2FA during their
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:689]:
```suggestion
In this case the user must configure 2FA during their
```

File: source/adminguide/accounts.rst [hunk 24]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+first login into CloudStack.
+
+The user's first login page to configure 2FA looks like the below,
+
+.. image:: /_static/images/configure-2fa-at-login-page.png
+   :width: 400px
+   :align: center
+   :alt: Configure 2FA at login page
+
+For the existing users, admin can mandate 2FA using the 'updateUser' API with the parameter 'mandate2FA'.
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:699]:
```suggestion
For the existing users, the admin can mandate 2FA using the 'updateUser' API with the parameter 'mandate2FA'.
```

File: source/adminguide/accounts.rst [hunk 25]
```diff
@@ -627,3 +627,80 @@ The following global configuration should be configured:
...
+
+Admin can also disable 2FA for users using the action button as shown below,
```
Comment (NuxRo) [inline source/adminguide/accounts.rst:701]:
```suggestion
The admin can also disable 2FA for a user using the action button as shown below.
```

**Review by NuxRo — CHANGES_REQUESTED** [review]

**Comment by harikrishna-patnala** [comment]
https://qa.cloudstack.cloud/builds/docs-build/pr/293/adminguide/accounts.html#using-two-factor-authentication-for-users

**Review by NuxRo — APPROVED** [review]
lgtm

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Comment by harikrishna-patnala** [comment]
@blueorangutan docbuild

**Review by rohityadavcloud — APPROVED** [review]

**Pull request merged**